src/spaces/homspace.jl

```diff
@@ -134,3 +140,119 @@ function compose(W::HomSpace{S}, V::HomSpace{S}) where {S}
     domain(W) == codomain(V) || throw(SpaceMismatch("$(domain(W)) ≠ $(codomain(V))"))
     return HomSpace(codomain(W), domain(V))
 end
+
+# Block and fusion tree ranges: structure information for building tensors
+#--------------------------------------------------------------------------
+struct FusionBlockStructure{I,F₁,F₂}
+    totaldim::Int
+    blockstructure::SectorDict{I,Tuple{Tuple{Int,Int},UnitRange{Int}}}
+    fusiontreelist::Vector{Tuple{F₁,F₂}}
+    fusiontreestructure::Vector{Tuple{Tuple{Int,Int},Tuple{Int,Int},Int}}
+    fusiontreeindices::FusionTreeDict{Tuple{F₁,F₂},Int}
+end
+
+abstract type CacheStyle end
+struct NoCache <: CacheStyle end
+struct TaskLocalCache{D<:AbstractDict} <: CacheStyle end
+struct GlobalLRUCache <: CacheStyle end
+
+function CacheStyle(I::Type{<:Sector})
+    return GlobalLRUCache()
+end
+
+fusionblockstructure(W::HomSpace) = fusionblockstructure(W, CacheStyle(sectortype(W)))
+
+function fusionblockstructure(W::HomSpace, ::NoCache)
+    codom = codomain(W)
+    dom = domain(W)
+    N₁ = length(codom)
+    N₂ = length(dom)
+    I = sectortype(W)
+    F₁ = fusiontreetype(I, N₁)
+    F₂ = fusiontreetype(I, N₂)
+
+    # output structure
+    blockstructure = SectorDict{I,Tuple{Tuple{Int,Int},UnitRange{Int}}}() # size, range
+    fusiontreelist = Vector{Tuple{F₁,F₂}}()
+    fusiontreestructure = Vector{Tuple{Tuple{Int,Int},Tuple{Int,Int},Int}}() # size, strides, offset
```

Comment:
I might be missing something here, but I find it a bit strange to store the `fusiontreestructure` sizes and strides as if it is a matrix. This means that within the `getindex(t, f1, f2)`, we first have to explicitly construct the `StridedView` as a matrix, and then reshape it into a `N1+N2` dimensional array. Can we not just immediately cache the size and strides of that resulting array?

Reply:
Probably, I know I tried as you suggested this before, but I found it conceptually much harder to compute and store that information in one go. I don't recall if there was a fundamental obstacle; I guess not. It is just that that information is much more entangled, as the strides for the "column" part depend on the "row" information. Datastructure wise, it then requires tuples of length `N1+N2`, whereas now only `F₁` and `F₂` are type parameters. I am wondering if there exists any case where the overhead of doing that final reshape of the `StridedView` matrix would really be significant, and would e.g. take more time than the global cache lookup itself.

Reply:
I bumped into this yesterday as I was trying to rewrite the `add_transform!` with a SparseMatrix. The thing is, if I need to have an additional reshape in obtaining the subblocks, it becomes a bit more involved as I then also need to simultaneously loop over the fusiontrees to have the uncoupled charges. I agree that the cached structure is absolutely conceptually hard to follow, but I would argue that "the strides for a matrix form of a block that still has to be reshaped" isn't that much better.

My main argument to changing this is however the following preliminary benchmark results I got with the SparseMatrix implementation:
```julia
V = SU2Space(i => 4 for i in 0:1//2:3)
t1 = rand(V^3);
p = ((3, 2, 1), ())
t2 = TensorKit.permute(t1, p);
```
```julia
julia> @benchmark add_permute!($t2, $t1, $p, $true, $false, $(NewBackend(StridedBLAS())))
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  34.656 μs … 66.047 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     35.047 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   35.910 μs ±  2.070 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▅█▆▂  ▅▇▄▂   ▁▁▁   ▁▁     ▃▂     ▁▂                         ▂
  █████▆███████████▅▇███▇▇▇▇██▇▆▅▆▆███▆▅▄▄▅▇▇▆▄▃▄▃▃▆█▅▅▄▃▁▃▃▆ █
  34.7 μs      Histogram: log(frequency) by time      44.8 μs <

 Memory estimate: 272 bytes, allocs estimate: 6.

julia> @benchmark add_permute!($t2, $t1, $p, $true, $false, $(StridedBLAS()))
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  204.842 μs …  19.483 ms  ┊ GC (min … max): 0.00% … 98.48%
 Time  (median):     216.778 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   234.075 μs ± 274.092 μs  ┊ GC (mean ± σ):  2.98% ±  2.99%

    █▇▅▃▃▃▄▆▆▃▁▁                                                ▁
  █████████████████▇▆▆▆▆▆▅▅▆▅▅▄▄▄▅▅▄▅▄▂▂▂▃▃▄▄▄▄▃▄▄▄▅▄▂▃▅▆▅▆▅▆▅▆ █
  205 μs        Histogram: log(frequency) by time        393 μs <

 Memory estimate: 81.59 KiB, allocs estimate: 1385.
```

I'll try to get it cleaned up a bit into a new branch, maybe you could run the entire benchmark suite again after?

Reply:
Wow, is that because of better multithreading? Or even in the single threaded case? Where is that speedup coming from?

So it seems that for some optimisations, having immediately the final StridedView field data for a given `(f1,f2)` is convenient, whereas for other optimisations (@maartenvd) the ideal data is the orignal `rowr` data for a given `f1` and separately `colr` for a given `f2` would be convenient? Whereas the current structure is a mix that jointly stores `rowr` and `colr` data for given `(f1,f2)`, which is optimal for neither case? :smile:

I will look into changing this tonight, and possibly having all of this available? We anyway need to first construct the old data before we can compute the `strides` data, so we can just as well cache everything.

Reply:
It's just single threaded, but with a whole lot less dictionary lookups

Reply:
That is amazing; I had such a large overhead for add_permute that I precalculated the exact offsets in the data, and circumvented most dict lookups! This version might be faster than what I spaghetti'd together.

Reply:
#167 is what I am working on. It could do with a large benchmark, but at the very least I feel it is quite elegant, and from my preliminary testing it is a lot faster. (Keep in mind that I disabled multithreading for now, which can also skew the results for smaller tensors)

My best guess is that really, having the offsets and everything precomputed in the caches, in vectors instead of dictionaries is just much less overhead, and additionally I think I managed to reduce the global cache lookups to 1 per transformation, instead of having to fetch that cache every single time a block needs to be extracted.

Reply:
If you want an actual tensor that I encountered during dmrg, you could look at the first contraction I did here https://github.com/maartenvd/MPSKitExperimental.jl/blob/main/examples/submult.ipynb The fastloop code will be broken, but I would be interested in knowing just how much faster you can get that particular contraction now